### PR TITLE
Us2 legenditems

### PR DESCRIPTION
**Pull request by hry625**
Functional wise the US2 are done.

### CODE REVIEW

**Comment by hry625** [comment]
截屏2022-10-14 00 10 00

**Comment by alrapal** [comment]
Hello! Good job! 

The combination of colours and shapes should help understanding the event types better. 

However, we should follow the PR template we decided previously. I will convert this PR as a draft for now. 

A few aspects that probably need improvement: 
- The Legend is very large and not filled which looks unfinished
- The last elements (SCC and SCS) are cut on the picture you sent, this should probably not happen, especially with a big legend box like that. 
- The legend button does not have a box attached to it which to me looks a bit odd (Like it was not supposed to be there). 

I know we have not talked about the look and the consistency of the design yet, as well as it was not part of the acceptance diagram, but we might want to think about it before we merge any code that touches the design 😄

**Comment by kam56** [comment]
Nice work!

I agree with Alex. The user story is implemented according to what was mentioned in the documentation so we need to have a more clear picture of what we want and what we mean by being DONE. 

We cannot merge now since we have not decided on a strategy for how we are gonna merge the PRs.

File: frontend/src/components/ColorLegend.svelte
Comment (RuthgerD) [inline frontend/src/components/ColorLegend.svelte:18]: Prefer functional programming, i.e.
```
styles.map(event => { id: event[0], size: 10, etc, etc })
```
Have you considered making this a computed value?
```
$: data = styles.map(...)
```

File: frontend/src/layout.ts
Comment (RuthgerD) [inline frontend/src/layout.ts:84]: Default case should be applied here (`defaultColor`) for the others as well

File: frontend/src/components/ColorLegend.svelte
Comment (hry625) [reply frontend/src/components/ColorLegend.svelte]: > Prefer functional programming, i.e.
> 
> ```
> styles.map(event => { id: event[0], size: 10, etc, etc })
> ```
> 
> Have you considered making this a computed value?
> 
> ```
> $: data = styles.map(...)
> ```

I have changed the code to functional programming.

File: frontend/src/layout.ts
Comment (hry625) [reply frontend/src/layout.ts:84]: I think the G6 lib come with default shape and color for the node and also the config.json file have all eiffle event coverd, so I did not implement that.

**Comment by hry625** [comment]
## Description: 
  * Add frontend/src/config.json to predefined colors and shapes for different eiffel events.
  * The preset colors are color blind friendly.
  * Modified frontend/src/ColorLegend and frontend/src/layout.ts to enable both the graph and legend to display the nodes as we designed.


## Useful notes:
* User could customerize the colors and shapes predefined in the frontend/src/config.json .
* The availiable shapes are circle, rect, triangle, diamond, eclipse according to [AntV/G6](https://github.com/antvis/G6/blob/master/packages/plugin/src/legend/index.ts) lib.

## Screenshots:
Preview of legend and nodes:
截屏2022-10-27 15 48 07

closes #6 & #2 & #1

File: frontend/src/layout.ts
Comment (kam56) [inline frontend/src/layout.ts:84]: There is an error regarding this line: layout.ts:84 `Uncaught(in promise) TypeError: Cannot read properties of undefined (reading 'Color')`

**Review by kam56 — CHANGES_REQUESTED** [review]
Looks very nice. However, the design does not follow the overall prototype:

- The width of the legend should be aligned with the filter options. 
- We do not have a title for the legend in the prototype. It looks better with the orange background though. But, the hover effect for the title, feels like the title is clickable. So I think the hover effect is not necessary. 
- We need to make room for the Graph Options as well on the same column as the legend/filter. So right now the legend has taken too much space. 

Please check the layout.js file, I have left a comment there.

File: frontend/src/layout.ts
Comment (kam56) [inline frontend/src/layout.ts:77]: Do we need these commented lines?

File: frontend/src/layout.ts
Comment (kam56) [reply frontend/src/layout.ts:84]: > There is an error regarding this line: layout.ts:84 `Uncaught(in promise) TypeError: Cannot read properties of undefined (reading 'Color')`

There was an issue with the event generator. So it works with database dump and looks very nice. The labels are also displayed atm. Since we have the legend with different shapes and colors, there is no need for labels for the nodes.

File: frontend/src/layout.ts
Comment (hry625) [reply frontend/src/layout.ts:78]: We keep it for now.

**Comment by kam56** [comment]
Nice. The hover effect and labels are removed. It looks very cool!

- The legend element's black background covers the graph, especially when it's in invisible mode. to minimize the covered space, I suggest that we set the main background to transparent. 

Screenshot 2022-10-27 at 23 20 58

**Comment by hry625** [comment]
> Nice. The hover effect and labels are removed. It looks very cool!
> 
> * The legend element's black background covers the graph, especially when it's in invisible mode. to minimize the covered space, I suggest that we set the main background to transparent.
> 
>Screenshot 2022-10-27 at 23 20 58

Since the US1 and US2 is only about legend and legend item. We can solve the blocking problem in the US14 branch.

File: frontend/src/config.json
Comment (alrapal) [inline frontend/src/config.json:3]: Minor detail but we maybe can change the name of the theme since it is not rainbow anymore

File: frontend/src/App.svelte
Comment (alrapal) [inline frontend/src/App.svelte:35]: Minor detail with the naming:
It is not rainbow anymore, maybe use something more generic since it is customisable also?

File: frontend/src/components/ColorLegend.svelte
Comment (alrapal) [inline frontend/src/components/ColorLegend.svelte:2]: I am not sure about moving all this here
It makes the overall architecture a bit confusing since it is here now that the graph is created whereas I would expect to have styles only. What is the purpose of the G6Graph.svelte file?

File: frontend/src/components/ColorLegend.svelte
Comment (alrapal) [inline frontend/src/components/ColorLegend.svelte:7]: The options are given as parameters in the app.svelte currently. Changing this pattern will impact other branches like #4 creating unnecessary conflicts in my opinion. 

(I don't understand why all this has been changed, but I am sure you had a reason! I just wonder if there is a way to have something that is less impacting9

File: frontend/src/config.txt
Comment (alrapal) [inline frontend/src/config.txt:2]: cf - comment on naming

**Review by alrapal — CHANGES_REQUESTED** [review]
Looks great!
- Some minor naming changes that I find couter intuitive 
- A big question mark on my side regarding the refactoring of the G6Graph.svelte and the ColorLegend.svelte and its impact on other ongoing branches

**Comment by alrapal** [comment]
> Looks great!
> 
> * Some minor naming changes that I find couter intuitive
> * A big question mark on my side regarding the refactoring of the G6Graph.svelte and the ColorLegend.svelte and its impact on other ongoing branches

After discussion with @hry625, things got way clearer. 
The legend is actually made of a G6 graph uploading the same colour theme as the main graph, through the same config file.  
My questions have then been answered and I don't see any conflicts anymore with #4 

We could consider renaming the component to Legend only since ColorLegend was a bit misleading (thought it was a component of the legend), but decision has to be evaluated first to avoid confusion and conflicts.

**Comment by alrapal** [comment]
Changed naming with @hry625 (to avoid having to change branch)

**Review by alrapal — APPROVED** [review]
Good for me!

**Comment by hry625** [comment]
> Nice. The hover effect and labels are removed. It looks very cool!
> 
> * The legend element's black background covers the graph, especially when it's in invisible mode. to minimize the covered space, I suggest that we set the main background to transparent.
> 
> Screenshot 2022-10-27 at 23 20 58

This will be fixed in the branch # 14

**Review by kam56 — APPROVED** [review]
As we discussed, the requested changes are planned to be added in the next related task, so this task is ready to be merged.

**Pull request merged**
